annotation/tests/test_tasks_services.py

```diff
@@ -585,3 +636,1697 @@ def test_filter_tasks_db_no_files_or_jobs(
         assert len(result[0]) == 2
         assert result[1] == expected_result[1]
         assert result[2] == expected_result[2]
+
+
+def test_create_tasks(mock_session: Mock):
+    tasks = [
+        {"user_id": 1, "file_id": 123, "pages": {1, 2}},
+        {"user_id": 2, "file_id": 456, "pages": {3}},
+    ]
+    job_id = 1
+    expected_user_ids = {1, 2}
+    with patch(
+        "annotation.tasks.services.update_files"
+    ) as mock_update_files, patch(
+        "annotation.tasks.services.update_user_overall_load"
+    ) as mock_update_user_overall_load:
+
+        create_tasks(mock_session, tasks, job_id)
+
+        mock_session.bulk_insert_mappings.assert_called_once()
+        mock_update_files.assert_called_once_with(mock_session, tasks, job_id)
+        mock_update_user_overall_load.assert_has_calls(
+            [call(mock_session, user_id) for user_id in expected_user_ids],
+            any_order=True,
+        )
```

Comment:
this and test_create_tasks_with_empty_tasks test has basically identical input data. Simply put it into parametrize

Reply:
Test will look like this:
@pytest.mark.parametrize(
    ("tasks", "job_id", "expected_user_ids", "should_call_update_user_overall_load"),
    (
        (
            [
                {"user_id": 1, "file_id": 123, "pages": {1, 2}},
                {"user_id": 2, "file_id": 456, "pages": {3}},
            ],
            1,
            {1, 2},
            True
        ),
        (
            [],
            1,
            set(),
            False
        )
    )
)
def test_create_tasks(mock_session: Mock, tasks: Dict[str, Union[int, set]], job_id:int, expected_user_ids,:set should_call_update_user_overall_load:bool):
    with patch(
        "annotation.tasks.services.update_files"
    ) as mock_update_files, patch(
        "annotation.tasks.services.update_user_overall_load"
    ) as mock_update_user_overall_load:
        services.create_tasks(mock_session, tasks, job_id)

        mock_session.bulk_insert_mappings.assert_called_once()
        mock_update_files.assert_called_once_with(mock_session, tasks, job_id)

        if should_call_update_user_overall_load:
            mock_update_user_overall_load.assert_has_calls(
                [call(mock_session, user_id) for user_id in expected_user_ids],
                any_order=True,
            )
        else:
            mock_update_user_overall_load.assert_not_called()

and this will add logic into test ( that is the reason why it is splited into two parts)

Reply:
I dont see the problem with assert_has_calls, where in first case you will have list with calls, in second you will have empty list.

P.S. use ```python and closing ``` to insert code blocks in comments. https://www.markdownguide.org/extended-syntax/#fenced-code-blocks